bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java

```diff
@@ -457,101 +488,121 @@ private boolean setupThing(Node node) {
+        }
 
-            ThingBuilder builder = editThing();
-            if (!result.location.equals(getThing().getLocation()) && !result.location.isBlank()) {
-                builder.withLocation(result.location);
-            }
+        ThingBuilder builder = editThing();
 
-            List<Channel> channelsToRemove = new ArrayList<>();
-            for (Channel channel : thing.getChannels()) {
-                if (!result.channels.containsKey(channel.getUID().getId())) {
-                    channelsToRemove.add(channel);
-                } else {
-                    result.channels.remove(channel.getUID().getId());
-                }
-            }
-            if (!channelsToRemove.isEmpty()) {
-                logger.trace(null, "Node {}. Removing {} channels", node.nodeId, channelsToRemove.size());
-                builder.withoutChannels(channelsToRemove);
+        // Update location if needed
+        if (!result.location.equals(getThing().getLocation()) && !result.location.isBlank()) {
+            builder.withLocation(result.location);
+        }
+
+        // Update channels
+        updateChannels(builder, result);
+
+        // Initialize state for channels and configuration
+        initializeChannelAndConfigState(node, result);
+
+        // Detect color capabilities
+        detectColorCapabilities(node);
+
+        return true;
+    }
+
+    private void updateChannels(ThingBuilder builder, ZwaveJSTypeGeneratorResult result) {
+        List<Channel> channelsToRemove = new ArrayList<>();
+        for (Channel channel : thing.getChannels()) {
+            if (!result.channels.containsKey(channel.getUID().getId())) {
+                channelsToRemove.add(channel);
+            } else {
+                result.channels.remove(channel.getUID().getId());
             }
-            if (!result.channels.isEmpty()) {
-                List<Channel> channels = new ArrayList<Channel>(result.channels.entrySet().stream()
-                        .sorted(Map.Entry.<String, Channel> comparingByKey()).map(m -> m.getValue()).toList());
-                logger.trace(null, "Node {}. Adding {} channels", node.nodeId, channels.size());
-                builder.withChannels(channels);
+        }
+        if (!channelsToRemove.isEmpty()) {
+            logger.trace(null, "Node {}. Removing {} channels", this.config.id, channelsToRemove.size());
+            builder.withoutChannels(channelsToRemove);
+        }
+        if (!result.channels.isEmpty()) {
+            List<Channel> channels = result.channels.entrySet().stream().sorted(Map.Entry.comparingByKey())
+                    .map(Map.Entry::getValue).collect(Collectors.toList());
+            logger.trace(null, "Node {}. Adding {} channels", this.config.id, channels.size());
+            builder.withChannels(channels);
+        }
+        if (!channelsToRemove.isEmpty() || !result.channels.isEmpty()) {
+            SemanticTag equipmentTag = getEquipmentTag(builder.build().getChannels());
```

Comment:
if you call builder.build here, perhaps it changes the builder state and interferes with the next builder.build call?

Reply:
First question:
-  First tests show this does not interfere.
Second question:
- result.channels only contain channels that are new. Existing not-changed channels are disregarded while they are essential for equipment tag determination.